# Spam Detector

In [1]:

# Import required libraries

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score


## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:

# Import the data

data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()


word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.


I predict the Random Forest Classifier Model to perform much better since the Random Forest Classifier Models are much more accurate when it comes to complex problems due to their design.


## Split the Data into Training and Testing Sets

In [3]:

# Create the labels set `y` and features DataFrame `X`

X = data.copy().drop(columns='spam')
y = data['spam']


In [4]:

# Check the balance of the labels variable (`y`) by using the `value_counts` function.

y.value_counts()


spam
0    2788
1    1813
Name: count, dtype: int64

In [5]:

# Split the data into X_train, X_test, y_train, y_test

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)
X_train.head()


word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
4576            0.00                0.0           0.00           0.0   
4401            0.00                0.0           0.00           0.0   
3707            0.17                0.0           0.17           0.0   
2362            0.00                0.0           0.00           0.0   
1537            0.00                0.0           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
4576           0.00             0.0               0.0                 0.0   
4401           0.00             0.0               0.0                 0.0   
3707           0.00             0.0               0.0                 0.0   
2362           0.00             0.0               0.0                 0.0   
1537           2.17             0.0               0.0                 0.0   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
4576              0.0             0.0  ...                   0.0        0.000   
4401              0.0             0.0  ...                   0.0        0.000   
3707              0.8             0.0  ...                   0.0        0.253   
2362              0.0             0.0  ...                   0.0        0.000   
1537              0.0             0.0  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
4576        0.131        0.000          0.0        0.000          0.0   
4401        0.000        0.000          0.0        0.000          0.0   
3707        0.168        0.084          0.0        0.024          0.0   
2362        0.000        0.000          0.0        0.000          0.0   
1537        0.000        0.000          0.0        0.000          0.0   

      capital_run_length_average  capital_run_length_longest  \
4576                       1.488                           5   
4401                       1.571                           5   
3707                       4.665                          81   
2362                       4.228                          53   
1537                       1.333                           5   

      capital_run_length_total  
4576                        64  
4401                        11  
3707                      1031  
2362                       148  
1537                        16  

[5 rows x 57 columns]

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [6]:

from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance

scaler = StandardScaler()


In [7]:

# Fit the Standard Scaler with the training data

_ = scaler.fit(X_train)


In [8]:

# Scale the training data

X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)


## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [9]:

# Train a Logistic Regression model and print the model score

from sklearn.linear_model import LogisticRegression


In [10]:

from sklearn.linear_model import LogisticRegression

lr_model = LogisticRegression(random_state=1).fit(X_train_scaled, y_train)
# lr_model.score(X_test_scaled,y_test)


In [11]:

# Make and save testing predictions with the saved logistic regression model using the test data

y_lr_predicted = lr_model.predict(X_test_scaled)

# Review the predictions

y_lr_predicted


array([0, 0, 1, ..., 0, 0, 1], dtype=int64)

In [12]:

# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.

print(f' ')
print(f'Logistic Regression Model')
print(f'Accuracy Score: ',accuracy_score(y_test, y_lr_predicted))
print(f' ')


 
Logistic Regression Model
Accuracy Score:  0.9287576020851434
 


## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [13]:

# Train a Random Forest Classifier model and print the model score

from sklearn.ensemble import RandomForestClassifier
rf_model = RandomForestClassifier(random_state=1).fit(X_train_scaled, y_train)
    

In [14]:

# Make and save testing predictions with the saved logistic regression model using the test data

y_rf_predicted = rf_model.predict(X_test_scaled)

# Review the predictions

y_rf_predicted


array([1, 1, 1, ..., 1, 0, 1], dtype=int64)

In [15]:

# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.

print(f' ')
print(f'Random Forest Classifier Model')
print(f'Accuracy Score: ',accuracy_score(y_test, y_rf_predicted))
print(f' ')


 
Random Forest Classifier Model
Accuracy Score:  0.9669852302345786
 


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.


The Random Forest Classifier Model performed better.

This is in line with my prediction and expectations as the Random Forest Classifier Models are better for complex problems due to their design.

The Random Forest Classifier Model has an accuracy score of 0.967 compared to the Logistic Regression Model at an accuracy of 0.928.

The difference was around 3.82% which under some circumstance may be a significant difference. 
